docs/spec_v6.md

```diff
@@ -0,0 +1,259 @@
+   routes feels like it's just supposed to pump Twirp's brand, and
+   provides no value back to users.
+ - Homophonous with "twerp": In some Very Serious settings (like
+   government websites), it's not okay that "twirp" sounds like
+   "twerp", which means something like "insignificant pest."
+
+## Overview
+
+The Twirp wire protocol is a simple RPC protocol based on HTTP and
+Protocol Buffers (proto). The protocol uses HTTP URLs to specify the
+RPC endpoints, and sends/receives proto messages as HTTP
+request/response bodies.
+
+To use Twirp, developers first define their APIs using proto files,
+then use Twirp tools to generate the client and the server libraries.
+The generated libraries implement the Twirp wire protocol, using the
+standard HTTP library provided by the programming language runtime or
+the operating system. Once the client and the server are implemented,
+the client can communicate with the server by making RPC calls.
+
+The Twirp wire protocol supports both binary and JSON encodings of
+proto messages, and works with any HTTP client and any HTTP version.
+
+## Overview
+
+The Twirp wire protocol is a simple RPC protocol based on HTTP and
+Protocol Buffers (proto). The protocol uses HTTP URLs to specify the
+RPC endpoints, and sends/receives proto messages as HTTP
+request/response bodies.
+
+To use Twirp, developers first define their APIs using proto files,
+then use Twirp tools to generate the client and the server libraries.
+The generated libraries implement the Twirp wire protocol, using the
+standard HTTP library provided by the programming language runtime or
+the operating system. Once the client and the server are implemented,
+the client can communicate with the server by making RPC calls.
+
+The Twirp wire protocol supports both binary and JSON encodings of
+proto messages, and works with any HTTP client and any HTTP version.
+
+### URLs
+
+In [ABNF syntax](https://tools.ietf.org/html/rfc5234), Twirp's URLs
+have the following format:
+
+```bnf
+URL ::= Base-URL "/" [ Package "." ] Service "/" Method
+```
+
+The Twirp wire protocol uses HTTP URLs to specify the RPC
+endpoints on the server for sending the requests. Such direct mapping
+makes the request routing simple and efficient. The Twirp URLs have
+the following components.
+
+* **Base-URL** is the virtual location of a Twirp API server, which is
+  typically published via API documentation or service discovery.
+  Currently, it should only contain URL `scheme` and `authority`. For
+  example, "https://example.com".
+* **Package** is the proto `package` name for an API, which is often
+  considered as an API version. For example,
+  `example.calendar.v1`. This component is omitted if the API
+  definition doesn't have a package name.
+* **Service** is the proto `service` name for an API. For example,
+  `CalendarService`.
+* **Method** is the proto `rpc` name for an API method. For example,
+  `CreateEvent`.
+
```

Comment:
Would it make sense to add an example? It may be useful, specially because Base-URL contains "/", and Package contains "." on it (same symbols used as separators). The example should make it visually clear that Twirp URLs need to be parsed from right to left (extract Method first, then Service, then optional Package, and the rest is the Base-URL).

Reply:
There is no need to parse the URL. The server registers a handler against a URL, and all it needs to do is URL lookup using a hash table.

Reply:
There's an example further down that I think should be enough.

I agree that URLs don't need to be parsed like that necessarily. For example, if the user mounts the Twirp server under a path prefix, we could say they should trim the path prefix from request URLs before handing the request off to Twirp in the Go implementation. In any case, this is an implementation-specific concern.

Reply:
Needing to parse right-to-left is worrisome. That will be hard to implement consistently and to explain concisely. It sounds like we're not quite heading in that direction .. but if we are, we should stop.

Reply:
@rhysh, certainly agreed, but I don't think that's what we're headed to. In Go, users would do something like this:
```go
mux.Handle("/userdefined" + HaberdasherPathPrefix, http.StripPrefix("/userdefined", svc))
```

Reply:
😅 I retire the sentence "need to be parsed from right to left". Sorry that was misleading. The URLs can be parsed in multiple ways, one of them is by starting from the right, but it could also be removing the Base-URL first and then using a regexp, etc. What I mean is that one example will make it easier to understand the things that need to be consider when implementing a Twirp router, or just to understand the routes themselves.